### PR TITLE
Support generic types

### PR DESCRIPTION
**Pull request by keradus**

### CODE REVIEW

**Comment by keradus** [comment]
ref #2841

tests for annotation itself for now are working, yet there are few more cases when utests are failing for some fixers

File: src/DocBlock/Annotation.php [hunk 1]
```diff
@@ -16,10 +16,16 @@
...
  *
  * @author Graham Campbell <graham@alt-three.com>
+ * @author Dariusz Rumiński <dariusz.ruminski@gmail.com>
  */
 class Annotation
 {
     /**
+     * @internal
+     */
+    const REGEX_TYPES = '(?<types>(?<type>(?:[\\\\@$]?\w+(?:\[\])?)|(?<generic>[\\\\@$]?\w+<(?:[\\\\@$]?\w+,\s*)?(?:(?&types)|(?&generic))>))(?:\|(?:(?&type)|(?&generic)))*)';
```
Comment (keradus) [inline src/DocBlock/Annotation.php]: i thought I saw `@this` somewhere, but it was `$this` :D indeed, could be simplified

File: src/DocBlock/Annotation.php [hunk 2]
```diff
@@ -147,7 +160,25 @@ public function getTag()
      */
     public function getTypes()
     {
-        return explode('|', $this->getTypesContent());
+        if (null === $this->types) {
+            $this->types = explode('|', $this->getTypesContent());
```
Comment (keradus) [inline src/DocBlock/Annotation.php]: nice catch, it's from old implementation when i was compering old behaviour vs new one

File: src/DocBlock/Annotation.php (hunk 1 above)
Comment (keradus) [inline src/DocBlock/Annotation.php]: astually, we have `@this` in `PhpdocReturnSelfReferenceFixerTest`.
i will add test to `AnnotationTest` as well

File: src/DocBlock/Annotation.php [hunk 3]
```diff
@@ -16,10 +16,39 @@
...
+    # <simple> is any non-array, non-generic, non-alternated type, eg `int` or `\Foo`
+    # <generic> is generic collection type, like `array<string, int>`, `Collection<Item>` and more complex like `Collection<int, \null|SubCollection<string>>`
+    # <type> is <simple>, <simple>[] or <generic> type, like `int`, `bool[]` or `Collection<ItemKey, ItemVal>`
+    # <types> is one or more types alternated via `|`, like `int|bool[]|Collection<ItemKey, ItemVal>`
+    (?<types>
+        (?<type>
+            (?:(?&simple)(?:\[\])?)
+            |
+            (?<generic>
+                (?<simple>
```
Comment (julienfalque) [inline src/DocBlock/Annotation.php]: I would swap the `simple` matching group here with the first reference to it a few lines above.

File: src/DocBlock/Annotation.php [hunk 4]
```diff
@@ -16,10 +16,39 @@
...
+                    [\\\\@$]?[\\\\\w]+
```
Comment (julienfalque) [inline src/DocBlock/Annotation.php]: No need to match `\` in the first character class as it will be matched by the second.

Comment (keradus) [inline src/DocBlock/Annotation.php]: good catch, originally I was checking only `\null`, then realised it could be `Foo\Bar`

File: src/DocBlock/Annotation.php (hunk 3 above)
Comment (keradus) [inline src/DocBlock/Annotation.php]: done

File: src/DocBlock/Annotation.php [hunk 5]
```diff
@@ -16,10 +16,46 @@
...
+    # <generic> is generic collection type, like `array<string, int>`, `Collection<Item>` and more complex like `Collection<int, \null|SubCollection<string>>`
+    # <type> is <simple>, <array> or <generic> type, like `int`, `bool[]` or `Collection<ItemKey, ItemVal>`
+    # <types> is one or more types alternated via `|`, like `int|bool[]|Collection<ItemKey, ItemVal>`
+    (?<types>
+        (?<type>
+            (?<array>
+                (?&scalar)\[\]
+            )
+            |
+            (?<scalar>
```
Comment (julienfalque) [inline src/DocBlock/Annotation.php]: Now I would swap `scalar` matching group with the first reference to it a few lines above 😁.

File: src/DocBlock/Annotation.php [hunk 6]
```diff
@@ -16,10 +16,46 @@
...
  */
 class Annotation
 {
     /**
+     * Regex to match any types, shall be used with `x` modifier.
+     *
+     * @internal
+     */
+    const REGEX_TYPES = '
+    # <scalar> is any non-array, non-generic, non-alternated type, eg `int` or `\Foo`
```
Comment (julienfalque) [inline src/DocBlock/Annotation.php]: `\Foo` isn't a scalar.

Comment (keradus) [inline src/DocBlock/Annotation.php]: it is. scalar, not primitvie

File: src/DocBlock/Annotation.php (hunk 5 above)
Comment (keradus) [inline src/DocBlock/Annotation.php]: i find this cleaner, as defined separatelly, nested defs like `<generic>` are not well readable

File: src/DocBlock/Annotation.php (hunk 6 above)
Comment (julienfalque) [inline src/DocBlock/Annotation.php]: I think you are confusing scalar and primitive:

>  PHP supports ten primitive types.
>
> - Four scalar types: boolean, integer, float (floating-point number, aka double), string.
> - Four compound types: array, object, callable, iterable.
> - And finally two special types: resource, NULL.

http://php.net/manual/en/language.types.intro.php

Comment (keradus) [inline src/DocBlock/Annotation.php]: then in php terminology it's neither scalar nor primitive...

Comment (julienfalque) [inline src/DocBlock/Annotation.php]: What about _&lt;simple&gt;_ then? Or _&lt;single&gt;_?

Comment (keradus) [inline src/DocBlock/Annotation.php]: `<type>` is kind of `<single>`, as there is no alternation. single could mean item of collection, like types vs type.
simple :+1:

Comment (SpacePossum) [inline src/DocBlock/Annotation.php]: :+1: for simple, does this apply to simple values as well like `false` in `@return false|array<int, false>` ?

File: tests/DocBlock/AnnotationTest.php [hunk 7]
```diff
@@ -239,7 +331,7 @@ public function provideTypesCases()
         return array(
             array(array('Foo', 'null'), array('Bar[]'), '     * @param Foo|null $foo', '     * @param Bar[] $foo'),
             array(array('false'), array('bool'), '*   @return            false', '*   @return            bool'),
-            array(array('RUNTIMEEEEeXCEPTION'), array('Throwable'), "\t@throws\t  \t RUNTIMEEEEeXCEPTION\t\t\t\t\t\t\t\n\n\n", "\t@throws\t  \t Throwable\t\t\t\t\t\t\t\n\n\n"),
+            array(array('RUNTIMEEEEeXCEPTION'), array('Throwable'), "* \t@throws\t  \t RUNTIMEEEEeXCEPTION\t\t\t\t\t\t\t\n\n\n", "* \t@throws\t  \t Throwable\t\t\t\t\t\t\t\n\n\n"),
```
Comment (SpacePossum) [inline tests/DocBlock/AnnotationTest.php]:
```
*\t@throws
*@throws
```

would still work? I'm not sure why this change is needed beside maybe adding the `*`

File: src/DocBlock/Annotation.php (hunk 6 above)
Comment (keradus) [inline src/DocBlock/Annotation.php]: yes. anything directly typehinted with php7 I would say

File: tests/DocBlock/AnnotationTest.php (hunk 7 above)
Comment (keradus) [inline tests/DocBlock/AnnotationTest.php]: yes, there was a missing `*`.
I will add tests for those.

Comment (keradus) [inline tests/DocBlock/AnnotationTest.php]: added

**Comment by keradus** [comment]
nullable types supported
